docs/BLOCK_DEVICE_LOCKING.md

```diff
@@ -98,3 +98,122 @@ Summarizing: it is recommended to take `LOCK_EX` BSD file locks when
 manipulating block devices in all tools that change file system block devices
 (`mkfs`, `fsck`, …) or partition tables (`fdisk`, `parted`, …), right after
 opening the node.
+
+# Example of Locking The Whole Disk
+
+The following is an example to leverage `libsystemd` infrastructure to get the whole disk of a block device and take the BSD lock on it.
+
+## Compile and Execute
+**Note that this example requires `libsystemd` version 251 or newer.**
+
+Place the code in a source file, e.g. `take_BSD_lock.c` and run the following commands:
+```
+$ gcc -o take_BSD_lock -lsystemd take_BSD_lock.c
+
+$ ./take_BSD_lock /dev/sda1
+Successfully taken the BSD lock: /dev/sda
+
+$ flock -x /dev/sda ./take_BSD_lock /dev/sda1
+Failed to take the BSD lock: /dev/sda
+Error: Resource temporarily unavailable
+```
+
+## Code
+```
+/* SPDX-License-Identifier: MIT-0 */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/file.h>
+#include <systemd/sd-device.h>
+#include <unistd.h>
+
+/**
+ * lock_whole_disk_from_devname
+ * @devname: devname of a block device, e.g., /dev/sda or /dev/sda1
+ * @open_mode: the mode to open the device, e.g., O_RDONLY|O_CLOEXEC|O_NONBLOCK|O_NOCTTY
+ * @lock_mode: the mode to take flock, e.g., LOCK_EX|LOCK_NB
+ *
+ * given the devname of a block device, take the BSD lock of the whole disk
+ *
+ * Returns: negative errno value on error, or non-negative fd if the lock was taken successfully.
+ **/
+int lock_whole_disk_from_devname(const char *devname, int open_mode, int lock_mode) {
+    __attribute__((cleanup(sd_device_unrefp))) sd_device *dev;
+    sd_device *whole_dev;
+    const char *whole_disk_devname, *subsystem, *devtype;
+    int r, fd = -1;
```

Comment:
`dev` must be initialized to `NULL`, otherwise if the call to `sd_device_new_from_devname` fails, `sd_device_unrefp` is called with garbage input.

--

It is nicer to move the declarations of the variables to the point where they are initialized. This is documentation code, and it's much easier to read code that doesn't require scanning up the page to figure out the variable type. (It also makes is *much* harder to use an uninitialized variable.)

For example, it's much nicer to declare `fd` at the point where it is assigned.

Above, add:
```
int closep(int *fd) {
    if (*fd >= 0)
        close(*fd);
}
```
and use `__attribute__((cleanup(closep))) fd = …`.